File: pypy/summands.py
"""Polymorphic summands: a base class plus three subclasses with different
`add` implementations. Summing over a random mix exercises a polymorphic
call site (the loop calls `s.add(total)` where `s` varies in type).
"""


class Summand:
    def __init__(self, value):
        self.value = value

    def add(self, total: int) -> int:
        raise NotImplementedError


class AddInt(Summand):
    # Add the stored int.
    def add(self, total: int) -> int:
        return total + self.value


class SquareInt(Summand):
    # Add the square of the stored int.
    def add(self, total: int) -> int:
        return total + self.value * self.value


class Identity(Summand):
    # Ignore the value; pass the running total through unchanged.
    def add(self, total: int) -> int:
        return total


KINDS = (AddInt, SquareInt, Identity)


def polysum(summands):
    total = 0
    for s in summands:
        total = s.add(total)
    return total

File: pypy/summands.js
"use strict";
// JS port of summands.py: a base class with three subclasses and polysum,
// which loops calling s.add(total) -- a polymorphic call site when the list
// holds a mix of types.

class Summand {
  constructor(value) {
    this.value = value;
  }
  add(total) {
    throw new Error("not implemented");
  }
}

class AddInt extends Summand {
  add(total) {
    return total + this.value;
  }
}

class SquareInt extends Summand {
  add(total) {
    return total + this.value * this.value;
  }
}

class Identity extends Summand {
  add(total) {
    return total;
  }
}

const KINDS = [AddInt, SquareInt, Identity];

function polysum(summands) {
  let total = 0;
  for (const s of summands) {
    total = s.add(total);
  }
  return total;
}

module.exports = { Summand, AddInt, SquareInt, Identity, KINDS, polysum };
